tests/integrations/server/test_stale_read.rs

```diff
@@ -0,0 +1,231 @@
+
+use futures::executor::block_on;
+use grpcio::{ChannelBuilder, Environment};
+use kvproto::kvrpcpb::{Context, GetResponse, Mutation, Op};
+use kvproto::metapb::Peer;
+use kvproto::tikvpb::TikvClient;
+use pd_client::PdClient;
+use raft::eraftpb::MessageType;
+use std::sync::Arc;
+use test_raftstore::*;
+use tikv_util::HandyRwLock;
+
+// A helpful wrapper to make the test logic clear
+struct PeerClient {
+    cli: TikvClient,
+    ctx: Context,
+}
+
+impl PeerClient {
+    fn new(cluster: &Cluster<ServerCluster>, region_id: u64, peer: Peer) -> PeerClient {
+        let cli = {
+            let env = Arc::new(Environment::new(1));
+            let channel =
+                ChannelBuilder::new(env).connect(&cluster.sim.rl().get_addr(peer.get_store_id()));
+            TikvClient::new(channel)
+        };
+        let ctx = {
+            let epoch = cluster.get_region_epoch(region_id);
+            let mut ctx = Context::default();
+            ctx.set_region_id(region_id);
+            ctx.set_peer(peer);
+            ctx.set_region_epoch(epoch);
+            ctx
+        };
+        PeerClient { cli, ctx }
+    }
+
+    fn kv_read(&self, key: Vec<u8>, ts: u64) -> GetResponse {
+        kv_read(&self.cli, self.ctx.clone(), key, ts)
+    }
+
+    fn must_kv_read_equal(&self, key: Vec<u8>, val: Vec<u8>, ts: u64) {
+        must_kv_read_equal(&self.cli, self.ctx.clone(), key, val, ts)
+    }
+
+    fn must_kv_write(&self, pd_client: &TestPdClient, kvs: Vec<Mutation>, pk: Vec<u8>) -> u64 {
+        must_kv_write(pd_client, &self.cli, self.ctx.clone(), kvs, pk)
+    }
+
+    fn must_kv_prewrite(&self, muts: Vec<Mutation>, pk: Vec<u8>, ts: u64) {
+        must_kv_prewrite(&self.cli, self.ctx.clone(), muts, pk, ts)
+    }
+
+    fn must_kv_commit(&self, keys: Vec<Vec<u8>>, start_ts: u64, commit_ts: u64) {
+        must_kv_commit(
+            &self.cli,
+            self.ctx.clone(),
+            keys,
+            start_ts,
+            commit_ts,
+            commit_ts,
+        )
+    }
+}
+
+// Testing how data replication clould effect stale read service
+#[test]
+fn test_stale_read_basic_flow_replicate() {
+    let mut cluster = new_server_cluster(0, 3);
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+
+    cluster.run();
+    cluster.sim.wl().start_resolved_ts_worker();
+
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+    let mut leader_client = PeerClient::new(&cluster, 1, new_peer(1, 1));
+    let mut follower_client2 = PeerClient::new(&cluster, 1, new_peer(2, 2));
+    // Set the `stale_read` flag
+    leader_client.ctx.set_stale_read(true);
+    follower_client2.ctx.set_stale_read(true);
+
+    // Filter read index request of follower 2, so it can't perform normal
+    // follower read
+    cluster.add_send_filter(CloneFilterFactory(
+        RegionPacketFilter::new(1, 2)
+            .direction(Direction::Send)
+            .msg_type(MessageType::MsgReadIndex),
+    ));
```

Comment:
Why this is necessary?

Reply:
See https://github.com/tikv/tikv/pull/10084#discussion_r622985902